File: combinatorics/resource/LCHP/problem_LTDT_code/bai_2/bai2.cpp
#include <iostream>
#include <vector>
#include <map>
using namespace std;

// Memoization cho tính toán nhanh
map<pair<int, int>, long long> memo;

// Hàm tính số phân hoạch của n với các phần không vượt quá max_part
long long partitionsWithMaxPart(int n, int max_part) {
    if (n == 0) return 1;
    if (n < 0 || max_part <= 0) return 0;
    
    pair<int, int> key = {n, max_part};
    if (memo.find(key) != memo.end()) {
        return memo[key];
    }
    
    long long result = partitionsWithMaxPart(n, max_part - 1);  // Không dùng max_part
    if (n >= max_part) {
        result += partitionsWithMaxPart(n - max_part, max_part);  // Dùng ít nhất 1 lần max_part
    }
    
    memo[key] = result;
    return result;
}

// Hàm tính số phân hoạch của n với phần tử lớn nhất CHÍNH XÁC là k
long long partitionsWithExactMax(int n, int k) {
    if (k > n) return 0;
    if (k == n) return 1;  // Chỉ có phân hoạch (n)
    
    // Số phân hoạch có max <= k trừ số phân hoạch có max <= k-1
    return partitionsWithMaxPart(n, k) - partitionsWithMaxPart(n, k - 1);
}

// Hàm tính p_k(n) - số phân hoạch của n với ĐÚNG k phần
long long partitionsWithExactParts(int n, int k) {
    if (k == 0) return (n == 0) ? 1 : 0;
    if (k > n) return 0;
    
    vector<vector<long long>> dp(n + 1, vector<long long>(k + 1, 0));
    dp[0][0] = 1;
    
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= min(i, k); j++) {
            for (int part = 1; part <= i; part++) {
                if (i - part >= 0 && j - 1 >= 0) {
                    dp[i][j] += dp[i - part][j - 1];
                }
            }
        }
    }
    
    return dp[n][k];
}

// Phương pháp sinh và đếm trực tiếp để kiểm tra
void generateAndCount(int n, int target_max, int current_sum, int current_max, 
                     vector<int>& current_partition, long long& count) {
    if (current_sum == n) {
        if (current_max == target_max) {
            count++;
            // In ra phân hoạch (tùy chọn)
            cout << "(";
            for (int i = 0; i < current_partition.size(); i++) {
                cout << current_partition[i];
                if (i < current_partition.size() - 1) cout << ", ";
            }
            cout << ") ";
        }
        return;
    }
    
    if (current_sum > n) return;
    
    int start = current_partition.empty() ? target_max : 
                min(target_max, current_partition.back());
    
    for (int i = start; i >= 1; i--) {
        if (current_sum + i <= n) {
            current_partition.push_back(i);
            generateAndCount(n, target_max, current_sum + i, 
                           max(current_max, i), current_partition, count);
            current_partition.pop_back();
        }
    }
}

int main() {
    int n, k;
    cout << "Nhập n và k: ";
    cin >> n >> k;
    
    if (n <= 0 || k <= 0) {
        cout << "n và k phải là số nguyên dương!\n";
        return 0;
    }
    
    cout << "Phân tích cho n = " << n << ", k = " << k << ":\n";
    cout << "==========================================\n";
    
    // Tính p_max(n, k) - số phân hoạch có phần tử lớn nhất là k
    long long p_max = partitionsWithExactMax(n, k);
    cout << "p_max(" << n << ", " << k << ") = " << p_max << "\n";
    
    // Tính p_k(n) - số phân hoạch có đúng k phần  
    long long p_k = partitionsWithExactParts(n, k);
    cout << "p_" << k << "(" << n << ") = " << p_k << "\n";
    
    cout << "\nSo sánh:\n";
    if (p_max > p_k) {
        cout << "p_max(" << n << ", " << k << ") > p_" << k << "(" << n << ")\n";
    } else if (p_max < p_k) {
        cout << "p_max(" << n << ", " << k << ") < p_" << k << "(" << n << ")\n";  
    } else {
        cout << "p_max(" << n << ", " << k << ") = p_" << k << "(" << n << ")\n";
    }
    
    // Liệt kê các phân hoạch có max = k (nếu số lượng không quá lớn)
    if (p_max <= 20) {
        cout << "\nCác phân hoạch có phần tử lớn nhất là " << k << ":\n";
        vector<int> partition;
        long long count = 0;
        generateAndCount(n, k, 0, 0, partition, count);
        cout << "\nTổng: " << count << " phân hoạch\n";
    }
    
    // Tính một số giá trị khác để so sánh
    cout << "\nBảng so sánh p_max và p_k cho các giá trị khác:\n";
    cout << "k\tp_max(n,k)\tp_k(n)\n";
    cout << "------------------------\n";
    
    for (int i = 1; i <= min(n, 10); i++) {
        long long pm = partitionsWithExactMax(n, i);
        long long pk = partitionsWithExactParts(n, i);
        cout << i << "\t" << pm << "\t\t" << pk << "\n";
    }
    
    return 0;
}

File: combinatorics/resource/LCHP/problem_LTDT_code/bai_2/bai2.py
# Bài 2: Đếm phân hoạch với phần tử lớn nhất là k

# Dictionary cho memoization
memo = {}

def partitions_with_max_part(n, max_part):
    """Tính số phân hoạch của n với các phần không vượt quá max_part"""
    if n == 0:
        return 1
    if n < 0 or max_part <= 0:
        return 0
    
    key = (n, max_part)
    if key in memo:
        return memo[key]
    
    # Không dùng max_part
    result = partitions_with_max_part(n, max_part - 1)
    
    # Dùng ít nhất 1 lần max_part
    if n >= max_part:
        result += partitions_with_max_part(n - max_part, max_part)
    
    memo[key] = result
    return result

def partitions_with_exact_max(n, k):
    """Tính số phân hoạch của n với phần tử lớn nhất CHÍNH XÁC là k"""
    if k > n:
        return 0
    if k == n:
        return 1  # Chỉ có phân hoạch (n)
    
    # Số phân hoạch có max <= k trừ số phân hoạch có max <= k-1
    return partitions_with_max_part(n, k) - partitions_with_max_part(n, k - 1)

def partitions_with_exact_parts(n, k):
    """Tính p_k(n) - số phân hoạch của n với ĐÚNG k phần"""
    if k == 0:
        return 1 if n == 0 else 0
    if k > n:
        return 0
    
    # Khởi tạo bảng DP
    dp = [[0 for _ in range(k + 1)] for _ in range(n + 1)]
    dp[0][0] = 1
    
    for i in range(1, n + 1):
        for j in range(1, min(i, k) + 1):
            for part in range(1, i + 1):
                if i - part >= 0 and j - 1 >= 0:
                    dp[i][j] += dp[i - part][j - 1]
    
    return dp[n][k]

def generate_and_count(n, target_max, current_sum=0, current_max=0, 
                      current_partition=None, count_ref=None):
    """Sinh và đếm trực tiếp các phân hoạch có max = target_max"""
    if current_partition is None:
        current_partition = []
    if count_ref is None:
        count_ref = [0]  # Sử dụng list để pass by reference
    
    if current_sum == n:
        if current_max == target_max:
            count_ref[0] += 1
            # In ra phân hoạch
            print(f"({', '.join(map(str, current_partition))})", end=" ")
        return
    
    if current_sum > n:
        return
    
    start = target_max if not current_partition else min(target_max, current_partition[-1])
    
    for i in range(start, 0, -1):
        if current_sum + i <= n:
            current_partition.append(i)
            generate_and_count(n, target_max, current_sum + i, 
                             max(current_max, i), current_partition, count_ref)
            current_partition.pop()

def main():
    print("Nhập n và k: ", end="")
    try:
        n, k = map(int, input().split())
    except:
        print("Vui lòng nhập hai số nguyên!")
        return
    
    if n <= 0 or k <= 0:
        print("n và k phải là số nguyên dương!")
        return
    
    print(f"Phân tích cho n = {n}, k = {k}:")
    print("=" * 40)
    
    # Clear memo cho tính toán mới
    memo.clear()
    
    # Tính p_max(n, k) - số phân hoạch có phần tử lớn nhất là k
    p_max = partitions_with_exact_max(n, k)
    print(f"p_max({n}, {k}) = {p_max}")
    
    # Tính p_k(n) - số phân hoạch có đúng k phần
    p_k = partitions_with_exact_parts(n, k)
    print(f"p_{k}({n}) = {p_k}")
    
    print("\nSo sánh:")
    if p_max > p_k:
        print(f"p_max({n}, {k}) > p_{k}({n})")
    elif p_max < p_k:
        print(f"p_max({n}, {k}) < p_{k}({n})")
    else:
        print(f"p_max({n}, {k}) = p_{k}({n})")
    
    # Liệt kê các phân hoạch có max = k (nếu số lượng không quá lớn)
    if p_max <= 20:
        print(f"\nCác phân hoạch có phần tử lớn nhất là {k}:")
        count_ref = [0]
        generate_and_count(n, k, count_ref=count_ref)
        print(f"\nTổng: {count_ref[0]} phân hoạch")
    
    # Tính một số giá trị khác để so sánh
    print("\nBảng so sánh p_max và p_k cho các giá trị khác:")
    print("k\tp_max(n,k)\tp_k(n)")
    print("-" * 24)
    
    for i in range(1, min(n, 10) + 1):
        pm = partitions_with_exact_max(n, i)
        pk = partitions_with_exact_parts(n, i)
        print(f"{i}\t{pm}\t\t{pk}")

if __name__ == "__main__":
    main()